app/backend/src/couchers/models.py

```diff
@@ -317,19 +317,63 @@ def __repr__(self):
 
 class UserSession(Base):
     """
-    Active session on the app, for auth
+    API keys/session cookies for the app
+
+    There are two types of sessions: long-lived, and short-lived. Long-lived are
+    like when you choose "remember this browser": they will be valid for a long
+    time without the user interacting with the site. Short-lived sessions on the
+    other hand get invalidated quickly if the user does not interact with the
+    site.
+
+    Long-lived tokens are valid from `created` until `expiry`.
+
+    Short-lived tokens expire after 2 hours if they are not used.
     """
 
     __tablename__ = "sessions"
     token = Column(String, primary_key=True)
 
     user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
 
-    # timezone should always be UTC
-    started = Column(DateTime(timezone=True), nullable=False, server_default=func.now())
+    # whether it's a long-lived or short-lived session
+    long_lived = Column(Boolean, nullable=False)
+
+    # the time at which the session was created
+    created = Column(DateTime(timezone=True), nullable=False, server_default=func.now())
+
+    # the expiry of the session: the session *cannot* be refreshed past this
+    expiry = Column(DateTime(timezone=True), nullable=False, server_default=func.now() + text("interval '7 days'"))
+
+    # the time at which the token was invalidated, allows users to delete sessions
+    deleted = Column(DateTime(timezone=True), nullable=True, default=None)
+
+    # the last time this session was used
+    last_seen = Column(DateTime(timezone=True), nullable=False, server_default=func.now())
+
+    # count of api calls made with this token/session (if we're updating last_seen, might as well update this too)
+    api_calls = Column(Integer, nullable=False, default=0)
+
+    # details of the browser, if available
+    # these are from the request creating the session, not used for anything else
+    ip_address = Column(String, nullable=True)
+    user_agent = Column(String, nullable=True)
 
     user = relationship("User", backref="sessions")
 
+    @hybrid_property
+    def is_valid(self):
+        """
+        It must have been created and not be expired or deleted.
+
+        Also, if it's a short lived token, it must have been used in the last 2 hours.
+        """
+        return (
+            (self.created <= func.now())
+            & (self.expiry >= func.now())
+            & (self.deleted == None)
+            & (self.long_lived | (func.now() - self.last_seen < text("interval '2 hours'")))
```

Comment:
Is the (bitwise?) or and and here intentional? I'm curious as to how this works!

Reply:
I believe the way this works is that the bitwise operators allow overloading, so sqlalchemy uses that to generate equivalent SQL, which coincides with the python logic, so it can be run at either end. This makes it a bunch faster to search for valid tokens.